### PR TITLE
fix issue 408

### PR DESCRIPTION
**Pull request by Innovationzk**
#408 send an empty packet when switching layer.

### CODE REVIEW

File: worker/src/RTC/SimulcastConsumer.cpp [hunk 1]
```diff
@@ -870,6 +870,45 @@ namespace RTC
 
 		if (isSyncPacket)
 		{
+			if (!lastSentRtpPacketHasMarker && shouldSwitchCurrentSpatialLayer)
```
Comment (ibc) [inline worker/src/RTC/SimulcastConsumer.cpp:873]: We always use `this->` for class instance members:

```suggestion
			if (!this->lastSentRtpPacketHasMarker && shouldSwitchCurrentSpatialLayer)
```

File: worker/src/RTC/SimulcastConsumer.cpp [hunk 2]
```diff
@@ -870,6 +870,45 @@ namespace RTC
...
+					  "failed to send empty packet [ssrc:%" PRIu32 ", seq:%" PRIu16 ", ts:%" PRIu32
+					  "] from original [ssrc:%" PRIu32 ", seq:%" PRIu16 ", ts:%" PRIu32 "]",
+					  emptyPacket->GetSsrc(),
+					  emptyPacket->GetSequenceNumber(),
+					  emptyPacket->GetTimestamp(),
+					  origSsrc,
+					  origSeq,
+					  origTimestamp);
+				}
+				delete emptyPacket;
```
Comment (ibc) [inline worker/src/RTC/SimulcastConsumer.cpp:909]:
```suggestion
				}

				delete emptyPacket;
```

**Review by ibc — CHANGES_REQUESTED** [review]
Two cosmetic required changes for now.

**Comment by ibc** [comment]
In theory it looks good (thanks for this) although I'm not sure about the implications of sending an empty packet without any RTP extension (abs-send-time, MID, etc). In theory nothing will happen. However, assuming you are using Chrome (so MID is in use for reception) to receive the simulcast based video stream, can you please dump the RTP packets sent by the Consumer and print here the dump of a normal RTP packet and also the empty packet with marker bit set?

In `Transport.cpp` in `OnConsumerSendRtpPacket()` method, just add this at the beginning:

```c++
packet->Dump();
```

@jmillan thoughs?

File: worker/src/RTC/SimulcastConsumer.cpp [hunk 3]
```diff
@@ -870,6 +870,45 @@ namespace RTC
...
+				header.ssrc           = packet->GetSsrc();
+
+				RTC::RtpPacket* emptyPacket =
+				  RTC::RtpPacket::Parse(reinterpret_cast<uint8_t*>(&header), sizeof(RTC::RtpPacket::Header));
+
+				if (this->rtpStream->ReceivePacket(emptyPacket))
+				{
+					this->listener->OnConsumerSendRtpPacket(this, emptyPacket);
+
+					this->rtpSeqManager.Offset(1);
```
Comment (jmillan) [inline worker/src/RTC/SimulcastConsumer.cpp:893]:
```suggestion
					this->listener->OnConsumerSendRtpPacket(this, emptyPacket);
					this->rtpSeqManager.Offset(1);
```

File: worker/include/RTC/SimulcastConsumer.hpp [hunk 4]
```diff
@@ -101,6 +101,7 @@ namespace RTC
 		std::vector<RTC::RtpStreamSend*> rtpStreams;
 		std::vector<RTC::RtpStream*> producerRtpStreams; // Indexed by spatial layer.
 		bool syncRequired{ false };
+		bool lastSentRtpPacketHasMarker{ true };
```
Comment (jmillan) [inline worker/include/RTC/SimulcastConsumer.hpp:104]: I would say this should be initialized to *false*

Comment (ibc) [reply worker/include/RTC/SimulcastConsumer.hpp:104]: If initialized to `true`, the first RTP packet the receiver endpoint will receive will be an empty packet. IMHO it's better to have it initialized to true so we avoid sending such a useless packet when the consumer sends its first packet.

Said that, I'm not sure if we should also unset this flag in other circumstances (when the consumer is paused/resumed, etc) or not.

Comment (penguinol) [reply worker/include/RTC/SimulcastConsumer.hpp:104]: > If initialized to `true`, the first RTP packet the receiver endpoint will receive will be an empty packet. IMHO it's better to have it initialized to true so we avoid sending such a useless packet when the consumer sends its first packet.
> 
> Said that, I'm not sure if we should also unset this flag in other circumstances (when the consumer is paused/resumed, etc) or not.

You are right, that's why we inited it as true. we will do some test tomorrow see what will happen when pasuing and resuming.

Comment (jmillan) [reply worker/include/RTC/SimulcastConsumer.hpp:104]: :+1:

**Comment by jmillan** [comment]
> In theory it looks good (thanks for this) although I'm not sure about the implications of sending an empty packet without any RTP extension (abs-send-time, MID, etc).

Good point. Since we are only sending this empty RTP packet *after* other RTP packets have already been sent, the RTP de-multiplexing on the receiver side should already be done and sending this packet with the correct SSRC and PT should be enough for the receiver to identify this packet.

**Comment by ibc** [comment]
I think that within `SimulcastConsumer::UserOnPaused()` and `SimulcastConsumer::UserOnTransportDisconnected()` the flag `this->lastSentRtpPacketHasMarker` should be set to `true` since later, when resumed or transport reconnected, it would be an scenario similar to when the Consumer is created so the flag should have the same value.

File: worker/src/RTC/SimulcastConsumer.cpp [hunk 5]
```diff
@@ -870,6 +870,45 @@ namespace RTC
...
+				header.padding        = 0u;
+				header.version        = 2u;
+				header.payloadType    = packet->GetPayloadType();
+				header.marker         = 1u;
+				header.sequenceNumber = seq;
+				header.timestamp      = this->rtpStream->GetMaxPacketTs();
+				header.ssrc           = packet->GetSsrc();
+
+				RTC::RtpPacket* emptyPacket =
+				  RTC::RtpPacket::Parse(reinterpret_cast<uint8_t*>(&header), sizeof(RTC::RtpPacket::Header));
```
Comment (ibc) [inline worker/src/RTC/SimulcastConsumer.cpp:887]: Note for us: we should provide a proper `RtpPacket` factory function. This code sucks (but this is the only that is possible for now).

To clarify, nothing to be done within this PR. This is a task we should deal with in the future.

Comment (penguinol) [reply worker/src/RTC/SimulcastConsumer.cpp:887]: Sad that the issus can still be reproduced after applying the patch.

We have two chrome browsers of the same version as video source. And another chrome broswer to watch them.
One of the video source had mosics when switching layers, but the other did't.

Comment (ibc) [reply worker/src/RTC/SimulcastConsumer.cpp:887]: And what is the behavior in the same scenario if you don't apply this patch?

**Pull request closed without merging**